### ISSUE TITLE
Packaging for via Flatpack/Flathub?

### ISSUE THREAD

**Issue by SchoolGuy** [umlet/umlet#640]
I am using this via the ZIP for ages now, since Java packaging is natively hard on most Linux Distros (I mean bundling stuff in a DEB/RPM is just pure evil, although common practice). I would love if we could instead just package a Flatpak and distribute it on Flathub.

I never did this myself but if the developers don't want to spend the time but would allow it, I would offer my help and time to maintain the Flatpak.

**Comment by afdia** [comment]
The only official Umlet distributions are those from www.umlet.com, Eclipse Marketplace and VSCode Marketplace.

Everything else (like the packages from some Linux distributions) is maintained by other people.
We also don't have any experience with Flatpak, but if you want to maintain a package there it would be nice to have (but I would suggest that you mention somewhere that the distribution is not managed by the umlet team)

By the way we are experimenting with https://github.com/fvarrui/JavaPackager to package umlet+JRE to avoid a dependency to a Java installation (in addition to our zip package).
There is a branch https://github.com/umlet/umlet/tree/packaging_jre where we successfully built a Windows Umlet versions which includes the JRE (with an overhead of ~40mb). To finalize the effort we would have to adapt our maven build to include a  section which downloads a version of adoptjdk for every OS and create a bundle for every OS (Linux, Windows, Macos)

**Comment by SchoolGuy** [comment]
Yes I agree with your points but from the perspective of the user it is easier having to type in `umlet` or have a start menu entry, then having a ZIP file "somewhere" on your system and the need to check for updates yourself. A package manager will always upgrade the software and you profit from updates. Because your bundled JRE will never recieve the latest Security and Stability updates from Upstream. A user will never download your Umlet weekly just to get the latest rebuild of your Bundle. Anyway...

I would consider your message an allowance to create such a package and maintain it for you. This is great news for me and will make my life much easier. Hopefully I can get you some users through the flatpack. Let's see. I will update this issue for my progress. I am finishing my apprenticeship currently, so everything is a littlebit slower as I am focusing on this currently...

**Comment by SchoolGuy** [comment]
P.S.: If possible I would love if you could assign me to this issue so I see it in my assigned issues on GH... :)

**Comment by afdia** [comment]
Ok thanks, I have assigned the issue to you.

**Comment by zocker-160** [comment]
@SchoolGuy any update on this? I would be interested in a Flatoak and willing to help you to get it to Flathub.

**Comment by SchoolGuy** [comment]
@zocker-160 I was sidetracked. Sorry. I completely forgot about this because my existing installation worked like a charm. I will start this today. :)

**Comment by SchoolGuy** [comment]
Okay the process is fairly straightforward it seems. It is documented here: https://github.com/flathub/flathub/wiki/App-Submission

**Comment by SchoolGuy** [comment]
Okay even though the process is straightforward, the execution is not it seems. The reason for this is quite satisfying actually: Flathub denies network access during the build time. Thus we have to work around this.

An example how this could look in the end is the following flathub entry: https://github.com/flathub/es.estoes.wallpaperDownloader

An explanation of the hows and whys can be found here: https://github.com/flatpak/flatpak-builder/issues/58#issuecomment-781508777

I will try to drive this in my fork and once I have an idea of the execution, I will try to open a WIP PR against all required repositories.

**Comment by zocker-160** [comment]
@SchoolGuy ah cool to hear from you, I personally already have experience with publishing on Flathub since I have helped other devs to do so.

My question was more like if I can help you in any way with this.

**Comment by SchoolGuy** [comment]
@zocker-160 Ahhh thanks a lot but I believe the Problem is Maven wanting to build something online, rather then offline. I have encountered this already with the openSUSE Buildservice already and solved it in the past. So it just takes time for me.

**Comment by zocker-160** [comment]
@SchoolGuy I created a working Flatpak using the prebuilt jar file

```yml
app-id: com.umlet.Umlet
runtime: org.freedesktop.Platform
runtime-version: "21.08"
sdk: org.freedesktop.Sdk
sdk-extensions:
  - org.freedesktop.Sdk.Extension.openjdk17

rename-icon: umlet_logo
rename-desktop-file: umlet.desktop
command: umlet

finish-args:
    - --env=PATH=/app/jre/bin:/usr/bin:/app/bin
    - --env=UMLET_HOME=/app/lib/umlet
    - --share=ipc
    - --socket=wayland
    - --socket=fallback-x11
    - --device=dri
    - --filesystem=home

modules:
    - name: openjdk
      buildsystem: simple
      build-commands:
        - /usr/lib/sdk/openjdk17/install.sh

    - name: Umlet
      buildsystem: simple
      build-commands:
        - install -D -m755 umlet.jar -t /app/lib/umlet
        - cp -a  {img,lib,palettes,custom_elements} /app/lib/umlet
        - install -D -m755 umlet.sh /app/bin/umlet
        # fix start command
        - sed -i -e 's/Exec=java -Dsun.java2d.xrender=f -jar umlet.jar/Exec=umlet/g' umlet.desktop
        - install -D -m644 umlet.desktop -t /app/share/applications
        - install -D -m644 img/umlet_logo.png -t /app/share/icons/hicolor/512x512/apps
#      post-install:
        # FIXME: create metainfo file
#        - install -D -m644 metainfo.xml -t /app/share/metainfo
      sources:
        - type: archive
          url: https://www.umlet.com/download/umlet_15_0/umlet-standalone-15.0.0.zip
          sha256: 81dbe1a981b2ac5b90861ae4176eb05dbdd340b4422e6e7dfee4b14cf9877401
```

**Comment by SchoolGuy** [comment]
@zocker-160 Credit is due where credit's due. Feel free to open a PR for this. I will try to improve that later on where we build this from source on Flathub. I will create an entry for Flathub once your PR is merged!

**Comment by agrrr3** [comment]
i think @zocker-160 never did the PR, so i am trying to make this work for myself currently.

currently one stumbling block is that AWT does not support wayland (so the flatpak yaml should use x11 socket), i hope that might change in the future https://groups.google.com/g/openjdk-discuss/c/SFRva5NsZ5c

**Comment by zocker-160** [comment]
@agrrr3 the reason for that was this issue here:
https://github.com/flatpak/flatpak/issues/4122

The font looks terrible to me, and I did not want to submit something that looks completely broken.

**Comment by agrrr3** [comment]
> The font looks terrible for me, and I did not want to submit something that looks completely broken.

What i see here i would not consider broken. but i am also not deep into fonts.

@zocker-160  you got newer answer in the linked discussion
> > I think the main issue is, that the Flatpak cannot access the fonts inside `/usr/share/fonts` and `/usr/local/share/fonts`
> 
> Your host fonts are mounted at `/run/host/fonts/` which should be loaded with right configuration (`/etc/fonts/conf.d/50- 50-flatpak.conf`), but, i'm not an expert with fontconfig.

in another thread (when libreoffice fixed their problems by also including /run/host/fonts in their c code) someone stated "The problem is possibly that this application bundles fontconfig which is broken. Just a thought though."

also i am not sure what the "correct" place for configuration is
https://github.com/flatpak/flatpak/issues/924#issuecomment-1214452211

also somebody said exposing the fontconfig directory via
`flatpak override --user --filesystem=xdg-config/fontconfig:ro` worked for them

i dont know where java/umlet looks for fonts, maybe we can include the ones used by flatpak there(?)

**Comment by agrrr3** [comment]
the flatpak shows in my environment (entered by  `flatpak run --command=bash --user --devel com.umlet.Umlet`)
```XDG_CONFIG_HOME=/home/agrrr3/.var/app/com.umlet.Umlet/config
XDG_CURRENT_DESKTOP=GNOME
XDG_DATA_DIRS=/app/share:/usr/share:/usr/share/runtime/share:/run/host/user-share:/run/host/share
```

according to XDG spec
`Config goes in XDG_CONFIG_HOME/fontconfig/fonts.conf or XDG_CONFIG_HOME/fontconfig/conf.d/, fonts are stored in XDG_DATA_HOME/fonts`

`fc-list` shows a lot of fonts under /run/host/fonts/ and some under /usr/share/fonts/

**Comment by afdia** [comment]
I did some experiments with your buildscript few weeks ago and was able to build a flatpak version of Umlet. My only issue was that the umlet_logo unintentionally had 1px more width than height (which was  the reason for https://github.com/umlet/umlet/commit/cccf693a2b602c88f74cdf0c796c3f9a7cbaa4b3 where I fixed that problem for the next release)

I don't really remember the font looking that bad, so I'm not sure if it's an isolated issue or a common one (maybe you could post a screenshot for comparison?)

In general I think a flatpak build would be nice for Linux users. If you could submit new versions to Flathub, it would be great.
Maybe just add a note to the description like "This is an unofficial build based on the zip-file releases from www.umlet.com" to let people know that we cannot really provide support for it but that it should be save to use because you just repackage the official zip files.
Maybe we will someday provide official builds to Flathub but in the meantime this would be our preferred way.

About the PR: If @zocker-160 provides the Flathub builds, it would be probably be better to put the buildscript into a own repository on his side to avoid confusion

**Comment by afdia** [comment]
I just tried it again by removing the logo reference (because in the v15 zip it's still wrong) and executing:
```
flatpak install org.freedesktop.Platform/x86_64/21.08 org.freedesktop.Sdk/x86_64/21.08 org.freedesktop.Sdk.Extension.openjdk17/x86_64/21.08
flatpak-builder --user --install --force-clean build-dir com.umlet.Umlet.yml
flatpak run com.umlet.Umlet
```
using the above mentioned script

The flatpak version is left, the zip file version is right:
![grafik](https://user-images.githubusercontent.com/8770707/202642619-a7d81188-48ae-45ad-96a5-2f53d407c850.png)

The font is not a perfect match but not really ugly IMO

**Comment by agrrr3** [comment]
> OT opinion: I hope that this font problems gets fixed on Flatpak's side, because it is an absurd amateur issue that packaging something in a specific format completely changes the appearance of the application.

ignoring the state of font configuration on linux (because i dont know enough about this) I find such an issue is neither surprising nor easy to solve in the general case.
e.g. flatpak is coming from the red hat side of things and i am mostly on debian and ubuntu systems and we have all that different window system and rendering toolkit differences. also java of course is also doing its own thing. so combinatorical complexity is quite high. packaging for a specific distro one has less moving targets (because the packager knows the default configs and the major versions of software).

an easy fix for having a nice font would be including one in the flatpak.

i am not sure what your expectation exactly is, but i guess: using the native fonts you configured in your settings.

other projects may want their application to look exactly like they ship it; i.e. completely ignoring fonts config

so maybe there is no one-size fits all. flatpak exposes the system fonts by default, so one can use those in a flatpak. 

maybe the java extension(?) should/could do something differently.

@zocker-160 also the user can override the configuration per application at least.  the formatting in my previous post cut off the most important variable for that
`XDG_CONFIG_HOME=/home/axel-testing/.var/app/com.umlet.Umlet/config`

linking your fontconfig directory from there could work (can you check your expected config works (or shows up/is accessible in the flatpak (via shell))

**Comment by zocker-160** [comment]
[small rant] @agrrr3 I have to disagree, solving this issue should be trivial:

You give the packager / user a simple switch to either use shipped fonts (to make the application look the same everywhere) or use host fonts by exposing  `/usr/share/fonts` and `/usr/local/share/fonts` to the flatpak.

But no instead flatpak developers in their ultimate wisdom decided to map it to `/run/host/fonts/` to make it completely incompatible with literally every application out there that does not ship flatpak specific fixes.

I also cannot force map it, because `/usr` [is blacklisted](https://docs.flatpak.org/en/latest/sandbox-permissions.html#filesystem-access). **facepalm**

[/small rant]

@afdia 

> In general I think a flatpak build would be nice for Linux users. If you could submit new versions to Flathub, it would be great.
Maybe just add a note to the description like "This is an unofficial build based on the zip-file releases from [www.umlet.com](http://www.umlet.com/)" to let people know that we cannot really provide support for it but that it should be save to use because you just repackage the official zip files.

Agreed, that makes sense to me.

There are two things that we need to discuss before I submit it:

1. FlatHub requires an appstream / metainfo file which is used to generate the page on FlatHub.org. It is preferred to have that available on the original repository, so that the maintainer can add changelogs, screenshots, description etc.
I need to know if you want to do that or if I should create my own independently (until maybe you want to offer it officially)

2. I need to know, who wants to be (co-)maintainer, changing that after submitting is always a bit cumbersome, so it would make sense to give everybody write access who currently or in the future might need it.

**Comment by afdia** [comment]
About 1: Correct me if I'm wrong, but I guess you will create a separate repository for the buildfile and I think that would be the best place for such a metainfo file, because then everything flatpak/flathub specific is in one place. If there is a new Umlet release, you can update the download zip path + hash + metainfo all at once and we don't have to coordinate between repositories.

About 2: I guess this question was targeted at other people in this thread, because if we someday decide to provide an official build, we would have to create a flathub ticket anyway in order to transfer ownership of the Umlet entry. If it helps and has no additional requirements you could already add me as co-maintainer, even tough I wouldn't really do any maintenance tasks.

**Comment by agrrr3** [comment]
> give the packager / user a simple switch to either use shipped fonts (to make the application look the same everywhere) or use host fonts by exposing `/usr/share/fonts` and `/usr/local/share/fonts` to the flatpak.
> 
a user based switch exposing configuration and fonts makes sense to me, maybe you could add a feature request to flatpak(?)

> ... flatpak developers ... decided to map it to `/run/host/fonts/` to make it completely incompatible with literally every application out there that does not ship flatpak specific fixes.
> 

in the original issue https://github.com/flatpak/flatpak/issues/1563 you'll find rationale for not directly using fontconfig files (AFAIU mostly a) that the config format is different depending on the version of the used  fontconfig binary and b) the configuration could be arbitrarily complex and one would need to basically drop any sandboxing to have this work in the general case)

**Comment by zocker-160** [comment]
> in the original issue https://github.com/flatpak/flatpak/issues/1563 you'll find rationale for not directly using fontconfig files

Yep I have read that thread and IMO and I think that both a) and b) are not really dealbreakers, because in case of an issue, the user can just fully switch to the fonts provided in the sandbox.

It would make sense to say, that exporting the host fonts by default is bad, but blocking that option completely by blacklisting `/usr` is what angers me.

**Comment by agrrr3** [comment]
> Yep I have read that thread and IMO and I think that both a) and b) are not really dealbreakers, because in case of an issue, the user can just fully switch to the fonts provided in the sandbox.

only if such a feature would exist; so maybe request it?

> but blocking that option completely by blacklisting `/usr` is what angers me.

there is probably a different rationale for blocking those system directories (my uninformed guess would be sandboxing issues in the general case). i would guess for read-only access to fonts it should be OK, at least after asking the user explicitly about it on install (which i think is not possible/maybe not planned at all).

**Comment by zocker-160** [comment]
Submission was accepted, Umlet is now on FlatHub!

**Comment by afdia** [comment]
That's great to hear, thank you for your effort! I guess I can close this ticket for now
